### PR TITLE
Opensource code for Deep Transformer with Latent Depth

### PR DESCRIPTION
**Pull request by xianxl**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Opensource code for Deep Transformer with Latent Depth (https://arxiv.org/pdf/2009.13102.pdf).

New features and design choices made:

- New feature: allow non-residual block to be weighted by sample z (generated per batch) instead of `x = residual + x`.
- Design choice: move  `x = residual + x` in transformer_layer.py into a function where the subclass (with latent depth) could overwrite it to `x = residual + z*x`.

- New feature: allow TransformerEncoder or TransformerDecoder to have additional logits parameters which will generate the samples z.
- Design choice: added subclass LatentTransformerEncoder and LatentTransformerDecoder, which has additional attributes for the logits parameters, and instantiate the corresponding LatentTransformerEncoderLayer and LatentTransformerDecoderLayer.

- New feature: allow multilingual_translation task to train with latent depth (results in the paper).  
- Design choice: 
  - added additional arguments in the multilingual_translation task. 
  - added option for multilingual_transformer to use LatentTransformerEncoder and LatentTransformerDecoder besides standard TransformerEncoder.
  - added option in multilingual_translation task's `train_step` to generate the samples z and compute the KL (and sparsity) loss per batch.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: examples/latent_depth/src/models/latent_multilingual_transformer.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from fairseq.models import (
```
Comment (myleott) [inline examples/latent_depth/src/models/latent_multilingual_transformer.py:1]: add copyright header

File: examples/latent_depth/src/multilingual_translation_latent_depth.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
+from fairseq.tasks import register_task
```
Comment (myleott) [inline examples/latent_depth/src/multilingual_translation_latent_depth.py:1]: add copyright header
